File: javascript/MergeSort.js
//ES6 syntax
//create a function to merge the arrays first
const merge = (arr1, arr2) => {
  let sorted = [];

  while (arr1.length && arr2.length) {
    if (arr1[0] < arr2[0]) sorted.push(arr1.shift());
    else sorted.push(arr2.shift());
  };

  return sorted.concat(arr1.slice().concat(arr2.slice()));
};

//recursively merge and sort
const mergeSort = arr => {
  if (arr.length <= 1) return arr;
  let mid = Math.floor(arr.length / 2),
      left = mergeSort(arr.slice(0, mid)),
      right = mergeSort(arr.slice(mid));

  return merge(left, right);
};

//test case
let unsortedArr = [1,23,4,5,6,32,2,4,12,3,5,25,3,5,1,23,4,5,6,32,2,4,12,3,5,25,3,5,1,23,4,5,6,32,2,4,12,3,5,25,3,5];

//output: [1, 1, 1, 2, 2, 2, 3, 3, 3, 3, 3, 3, 4, 4, 4, 4, 4, 4, 5, 5, 5, 5, 5, 5, 5, 5, 5, 6, 6, 6, 12, 12, 12, 23, 23, 23, 25, 25, 25, 32, 32, 32]
console.log(mergeSort(unsortedArr));

File: python/MergeSort.py
def merge(arr, L, R): 
  i = j = k = 0

  # Copy data to temp arrays L[] and R[] 
  while i < len(L) and j < len(R): 
      if L[i] < R[j]: 
          arr[k] = L[i] 
          i+= 1
      else: 
          arr[k] = R[j] 
          j+= 1
      k+= 1
    
  # Checking if any element was left 
  while i < len(L): 
      arr[k] = L[i] 
      i+= 1
      k+= 1
    
  while j < len(R): 
      arr[k] = R[j] 
      j+= 1
      k+= 1

def mergeSort(arr):
  if len(arr) > 1: 
    mid = len(arr)//2   # Finding the mid of the array 
    L = arr[:mid]       # Dividing the array elements into 2 halves  
    R = arr[mid:]    

    mergeSort(L) # Sorting the first half 
    mergeSort(R) # Sorting the second half
    merge(arr, L, R) # Merging the 2 halves
  
  
# Code to print the list 
def printList(arr): 
  for i in range(len(arr)):         
      print(arr[i], end =" ") 
  print() 
  
# driver code to test the above code 
if __name__ == '__main__': 
  arr = [1,23,4,5,6,32,2,4,12,3,5,25,3,5,1,23,4,5,6,32,2,4,12,3,5,25,3,5,1,23,4,5,6,32,2,4,12,3,5,25,3,5];  
  print ("Given array is", end ="\n")  
  printList(arr) 
  mergeSort(arr) 
  print("Sorted array is: ", end ="\n") 
  printList(arr) 